### ISSUE TITLE
about DUL KL-loss

### ISSUE THREAD

**Issue by swy755565055** [Ontheway361/dul-pytorch#9]
I have used this code for training. The arcface can converge normally, but the kl loss does not decrease. And the trained var-branch has no effect.

**Comment by JieChang-M** [comment]
I am the author of this paper, the KL loss shouldn't be overly minimized (this is why we set \lambda as 0.01 or 0.001). Practically, for Face Recognition task, the KL loss decrease first and then increase during the training, and at last, KL loss will fluctuates around a value. (MS1M + 30 epoch). 

If you set \lambda too large. KL loss will converge fast so that the softmax loss cannot be converge.

Of course, I think the \lambda is a hyper-parameter you need to select via cross-validation according to your own training set.

**Comment by CMACH508** [comment]
> I am the author of this paper, the KL loss shouldn't be overly minimized (this is why we set \lambda as 0.01 or 0.001). Practically, for Face Recognition task, the KL loss decrease first and then increase during the training, and at last, KL loss will fluctuates around a value. (MS1M + 30 epoch).
> 
> If you set \lambda too large. KL loss will converge fast so that the softmax loss cannot be converge.
> 
> Of course, I think the \lambda is a hyper-parameter you need to select via cross-validation according to your own training set.

After training this code, it is found that each component of var approaches the same value，and the var of all samples are similar. It seems that trained var-branch has no effect.  Do you know what might be the problem?

**Comment by YouWillLikeIt** [comment]
> I am the author of this paper, the KL loss shouldn't be overly minimized (this is why we set \lambda as 0.01 or 0.001). Practically, for Face Recognition task, the KL loss decrease first and then increase during the training, and at last, KL loss will fluctuates around a value. (MS1M + 30 epoch).
> 
> If you set \lambda too large. KL loss will converge fast so that the softmax loss cannot be converge.
> 
> Of course, I think the \lambda is a hyper-parameter you need to select via cross-validation according to your own training set.

After training this code, it is found that each component of var approaches the same value，and the var of all samples are similar. It seems that trained var-branch has no effect. Do you know what might be the problem?

**Comment by JieChang-M** [comment]
> > I am the author of this paper, the KL loss shouldn't be overly minimized (this is why we set \lambda as 0.01 or 0.001). Practically, for Face Recognition task, the KL loss decrease first and then increase during the training, and at last, KL loss will fluctuates around a value. (MS1M + 30 epoch).
> > If you set \lambda too large. KL loss will converge fast so that the softmax loss cannot be converge.
> > Of course, I think the \lambda is a hyper-parameter you need to select via cross-validation according to your own training set.
> 
> After training this code, it is found that each component of var approaches the same value，and the var of all samples are similar. It seems that trained var-branch has no effect. Do you know what might be the problem?

As Fig.6 illustrated in the paper, the uncertainty (we set uncertainty as the harmonic mean of the predicted variance) of most high-quality face images are indeed similar and small, wherever the uncertainty of few low-quality face images are pretty large. 

"Each component of var approaches the same value". It is true, we also observed this phenomenon. This is also happened in PFE. Actually, we set var branch as a single unit output fc layer in PFE and DUL, and then we found the performances didn't decrease or slightly decrease.

Email me (changjie@megvii.com) if you have more questions.

**Comment by niunai0906** [comment]
> I am the author of this paper, the KL loss shouldn't be overly minimized (this is why we set \lambda as 0.01 or 0.001). Practically, for Face Recognition task, the KL loss decrease first and then increase during the training, and at last, KL loss will fluctuates around a value. (MS1M + 30 epoch).
> If you set \lambda too large. KL loss will converge fast so that the softmax loss cannot be converge.
> Of course, I think the \lambda is a hyper-parameter you need to select via cross-validation according to your own training set.
> 
> After training this code, it is found that each component of var approaches the same value，and the var of all samples are similar. It seems that trained var-branch has no effect. Do you know what might be the problem?

No matter what image (noise or face) is input, the output value of VaR branch is the same. I have the same problem as you that trained var-branch of this code has no effect

**Comment by sakura811** [comment]
> > I am the author of this paper, the KL loss shouldn't be overly minimized (this is why we set \lambda as 0.01 or 0.001). Practically, for Face Recognition task, the KL loss decrease first and then increase during the training, and at last, KL loss will fluctuates around a value. (MS1M + 30 epoch).
> > If you set \lambda too large. KL loss will converge fast so that the softmax loss cannot be converge.
> > Of course, I think the \lambda is a hyper-parameter you need to select via cross-validation according to your own training set.
> > After training this code, it is found that each component of var approaches the same value，and the var of all samples are similar. It seems that trained var-branch has no effect. Do you know what might be the problem?
> 
> No matter what image (noise or face) is input, the output value of VaR branch is the same. I have the same problem as you that trained var-branch of this code has no effect

Hi, I also have the same problem with you, no matter what image (noise or normal face) is input, the ouput value of var is almost the same. 
                  normal _0000212_160.jpg 0.044065243267058296
GaussianBlur_5_1.5_0000212_160.jpg 0.044067524794118504
GaussianBlur_9_1.5_0000212_160.jpg 0.04406601733011649
GaussianBlur_9_3.0_0000212_160.jpg 0.044065361782127055

do you have solved this problem ?

**Comment by sakura811** [comment]
> > > I am the author of this paper, the KL loss shouldn't be overly minimized (this is why we set \lambda as 0.01 or 0.001). Practically, for Face Recognition task, the KL loss decrease first and then increase during the training, and at last, KL loss will fluctuates around a value. (MS1M + 30 epoch).
> > > If you set \lambda too large. KL loss will converge fast so that the softmax loss cannot be converge.
> > > Of course, I think the \lambda is a hyper-parameter you need to select via cross-validation according to your own training set.
> > 
> > 
> > After training this code, it is found that each component of var approaches the same value，and the var of all samples are similar. It seems that trained var-branch has no effect. Do you know what might be the problem?
> 
> As Fig.6 illustrated in the paper, the uncertainty (we set uncertainty as the harmonic mean of the predicted variance) of most high-quality face images are indeed similar and small, wherever the uncertainty of few low-quality face images are pretty large.
> 
> "Each component of var approaches the same value". It is true, we also observed this phenomenon. This is also happened in PFE. Actually, we set var branch as a single unit output fc layer in PFE and DUL, and then we found the performances didn't decrease or slightly decrease.
> 
> Email me ([changjie@megvii.com](mailto:changjie@megvii.com)) if you have more questions.

Hi, I also have the same problem, even if low-quality face images, the ouputs value of var are almost the same.
do you know what is the reason ? thank you.

**Comment by sakura811** [comment]
> > I am the author of this paper, the KL loss shouldn't be overly minimized (this is why we set \lambda as 0.01 or 0.001). Practically, for Face Recognition task, the KL loss decrease first and then increase during the training, and at last, KL loss will fluctuates around a value. (MS1M + 30 epoch).
> > If you set \lambda too large. KL loss will converge fast so that the softmax loss cannot be converge.
> > Of course, I think the \lambda is a hyper-parameter you need to select via cross-validation according to your own training set.
> > After training this code, it is found that each component of var approaches the same value，and the var of all samples are similar. It seems that trained var-branch has no effect. Do you know what might be the problem?
> 
> No matter what image (noise or face) is input, the output value of VaR branch is the same. I have the same problem as you that trained var-branch of this code has no effect

Hi, I also have the same problem, but I found I made mistakes on sigma and epsilon of s(i). maybe you can check them.
